File: src/typescript/delete-pet.step.ts
// src/typescript/delete-pet.step.ts
import { ApiRouteConfig, Handlers } from 'motia'
import { z } from 'zod'
import { TSStore } from './ts-store'

// Define path parameters schema for type safety
const pathParamsSchema = z.object({
  id: z.string().min(1, 'Pet ID is required'),
})

// Define event data schema for type safety
export const config: ApiRouteConfig = {
  type: 'api',
  name: 'TsDeletePet',
  path: '/ts/pets/:id',
  method: 'DELETE',
  emits: [],
  flows: ['TsPetManagement'],
}

export const handler: Handlers['TsDeletePet'] = async (
  req,
  { emit, logger }
) => {
  try {
    // Validate path parameters
    const { id } = pathParamsSchema.parse(req.pathParams)

    const deletedPet = TSStore.softDelete(id)

    if (!deletedPet) {
      return {
        status: 404,
        body: { message: 'Pet not found' },
      }
    }

    if (logger) {
      logger.info('🗑️ Pet soft deleted', {
        petId: deletedPet.id,
        name: deletedPet.name,
        purgeAt: new Date(deletedPet.purgeAt!).toISOString(),
      })
    }

    // Pet soft deleted successfully

    return {
      status: 202,
      body: {
        message: 'Pet scheduled for deletion',
        petId: deletedPet.id,
        purgeAt: deletedPet.purgeAt,
      },
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      return {
        status: 400,
        body: {
          message: 'Invalid pet ID',
          errors: error.errors,
        },
      }
    }

    // Handle unexpected errors
    if (logger) {
      logger.error('❌ Pet deletion failed', {
        error: error instanceof Error ? error.message : 'Unknown error',
      })
    }

    return {
      status: 500,
      body: { message: 'Internal server error' },
    }
  }
}

File: src/python/delete_pet_step.py
# src/python/delete_pet.step.py
config = { "type":"api", "name":"PyDeletePet", "path":"/py/pets/:id", "method":"DELETE", "emits": [], "flows": ["PyPetManagement"] }

async def handler(req, _ctx=None):
    try:
        import sys
        import os
        sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
        from src.services.pet_store import remove
    except ImportError:
        return {"status": 500, "body": {"message": "Import error"}}
    pid = req.get("pathParams", {}).get("id")
    ok = remove(pid)
    return {"status": 204, "body": {}} if ok else {"status": 404, "body": {"message":"Not found"}}
